src/data/vector/basic.lean

```diff
@@ -289,24 +289,61 @@ def mmap {m} [monad m] {α} {β : Type u} (f : α → m β) :
   do h' ← f a, t' ← mmap f v, pure (h' ::ᵥ t')
 | _ ⟨l, rfl⟩ := rfl
 
-/-- Define `C v` by induction on `v : vector α (n + 1)`, a vector of
-at least one element.
-This function has two arguments: `h0` handles the base case on `C nil`,
-and `hs` defines the inductive step using `∀ x : α, C v → C (x ::ᵥ v)`. -/
-@[elab_as_eliminator] def induction_on
-  {α : Type*} {n : ℕ}
-  {C : Π {n : ℕ}, vector α n → Sort*}
-  (v : vector α (n + 1))
-  (h0 : C nil)
-  (hs : ∀ {n : ℕ} {x : α} {w : vector α n}, C w → C (x ::ᵥ w)) :
+/-- Define `C v` by induction on `v : vector α n`.
+
+This function has two arguments: `h_nil` handles the base case on `C nil`,
+and `h_cons` defines the inductive step using `∀ x : α, C w → C (x ::ᵥ w)`. -/
+@[elab_as_eliminator] def induction_on {C : Π {n : ℕ}, vector α n → Sort*}
+  (v : vector α n)
+  (h_nil : C nil)
+  (h_cons : ∀ {n : ℕ} {x : α} {w : vector α n}, C w → C (x ::ᵥ w)) :
     C v :=
 begin
-  induction n with n hn,
-  { rw ←v.cons_head_tail,
-    convert hs h0 },
-  { rw ←v.cons_head_tail,
-    apply hs,
-    apply hn }
+  induction n with n ih generalizing v,
+  { rcases v with ⟨_|⟨-,-⟩,-|-⟩,
+    exact h_nil, },
+  { rcases v with ⟨_|⟨a,v⟩,_⟩,
+    cases v_property,
+    apply @h_cons n _ ⟨v, (add_left_inj 1).mp v_property⟩,
+    apply ih, }
+end
+
+/-- Define `C v w` by induction on a pair of vectors `v w : vector α n`. -/
+@[elab_as_eliminator] def induction_on₂ {C : Π {n}, vector α n → vector α n → Sort*}
+  (v w : vector α n)
+  (h_nil : C nil nil)
+  (h_cons : ∀ {n} {a b : α} {x y : vector α n}, C x y → C (a ::ᵥ x) (b ::ᵥ y)) : C v w :=
```

Comment:
What do you think about using different types for `v` and `w`?
```suggestion
@[elab_as_eliminator] def induction_on₂ {C : Π {n}, vector α n → vector β n → Sort*}
  (v : vector α n) (w : vector β n)
  (h_nil : C nil nil)
  (h_cons : ∀ {n a b} {x : vector α n} {y : vector β n}, C x y → C (a ::ᵥ x) (b ::ᵥ y)) : C v w :=
```
Similarly for the next lemma.
Otherwise LGTM.
bors d+